### PR TITLE
Adding one-time RPC to find unaliased / true dataset ID.

### PR DESCRIPTION
**Pull request by dhermes**
Also removing unnecessary functions that are no longer
needed since there is no need to muck with the dataset ID.

@tseaver This is really an exploratory PR, no need to review just yet.

@pcostell Can you check out the `_find_true_dataset_id` method in `gcloud/datastore/__init__.py` (feel free to ignore the rest).

Is this a dumb idea? Is there an actual API available to convert from the "human friendly" ID `foo` to the unaliased / prefixed `s~foo`?

### CODE REVIEW

**Comment by pcostell** [comment]
Pretty soon the datastore will normalize all types of keys, so you should never need to provide anything except the project id. Importantly, `_find_true_dataset_id` will only work with the current dataset which means it will break for users that store foreign keys.

In the near-future we don't want anyone to have to deal with the "s~", it should be purely an implementation detail of the Datastore. I think the client libraries should probably be written pretending the "s~" doesn't exist since it is cleaner that way (with the caveat that it is slightly broken atm).

**Comment by dhermes** [comment]
Thanks for the quick followup. That is great news! It will certainly be much cleaner if we can ignore `s~`. All the deleted code in this PR will still be able to be deleted even without `_find_true_dataset_id`.

How do you mean normalize? Currently it seems some normalization is happening, but that means if I send no dataset ID in the request and `foo` as the dataset ID in the URI, then I get back `s~foo` on the keys in the response. This still causes some issues.

**Comment by pcostell** [comment]
I believe the goal is that everywhere you will specify `foo` or an empty dataset and when you get the key back you will also only see `foo` in the dataset. 

@eddavisson can you confirm?

**Comment by eddavisson** [comment]
That's correct. Note that if you want to do a one-time RPC in the mean time, you can do a Lookup request and check both the found and missing fields in the response for the resolved dataset id (basically what you're doing in this pull request but it's a read-only RPC).

**Comment by dhermes** [comment]
I didn't think to use `missing`! Good call. With the `__key__` projection that'd be cheap enough, even if somehow the user had stored `Key('__Foo', 1)`.

**Comment by silvolu** [comment]
I think we should go forward with @eddavisson suggestion while waiting for the API updates to land.

**Comment by dhermes** [comment]
@pcostell RE: foreign keys. I'm not sure why that's an issue if the application is managing those keys.

I double checked and the backend does not change the foreign dataset ID after storing (seems it shouldn't, for privacy reasons). From 0525530ad7a76e408ca2e842b883d49c353ed56d:

```python
>>> from gcloud import datastore
>>> from gcloud.datastore import _implicit_environ
>>> from gcloud.datastore.entity import Entity
>>> from gcloud.datastore.key import Key
>>>
>>> datastore.set_default_connection()
>>> dataset_id = 'foo'
>>> datastore.set_default_dataset_id(dataset_id)
>>>
>>> e = Entity(key=Key('Foo', 1))
>>> e['bar'] = Key('Baz', 1, dataset_id='foreignappid')
>>> e.save()
>>>
>>> e_retrieved = datastore.get(e.key)
>>> e_retrieved['bar'].dataset_id
u'foreignappid'
>>>
>>> entity_pb, = _implicit_environ.CONNECTION.lookup(
...     dataset_id=dataset_id,
...     key_pbs=[e.key.to_protobuf()],
... )
>>> entity_pb.property
[<gcloud.datastore.datastore_v1_pb2.Property object at 0x7f4a5006d398>]
>>> print entity_pb.property[0].value.key_value
partition_id {
  dataset_id: "foreignappid"
}
path_element {
  kind: "Baz"
  id: 1
}

>>> e.key.delete()
```

FWIW I tried this both with a foreign dataset that my (user, not service account) OAuth 2.0 token had access to and one that the token did not.

**Comment by pcostell** [comment]
I believe we want to remove the notion of "s~" completely. So anywhere the user interacts with a dataset it should never have this extra prefix, including foreign keys. 

Note that this is not implemented, so it isn't finalized and definitely won't work yet.

File: gcloud/datastore/__init__.py [hunk 1]
```diff
@@ -68,6 +68,49 @@
...
+    :returns: The true / prefixed / un-aliased dataset ID.
+    """
+    if dataset_id.startswith('s~') or dataset_id.startswith('e~'):
+        return dataset_id
+
+    connection = connection or _implicit_environ.CONNECTION
+
+    # Create the bogus Key protobuf to be looked up and remove
+    # the dataset ID so the backend won't complain.
+    bogus_key_pb = Key('__Foo', 1, dataset_id=dataset_id).to_protobuf()
```
Comment (tseaver) [inline gcloud/datastore/__init__.py]: Should the `kind` be more indicative of purpose?  E.g., to help when looking at server-side logs if something goes amuck.

Comment (dhermes) [inline gcloud/datastore/__init__.py]: Good call. WDYT of `__MissingLookupKind`?

Comment (tseaver) [inline gcloud/datastore/__init__.py]: That name SGTM.

**Comment by dhermes** [comment]
Had to rebase after #548 and #550, so did the renames in each commit through the rebase.

File: gcloud/datastore/batch.py [hunk 2]
```diff
@@ -299,7 +299,6 @@ def _assign_entity_to_mutation(mutation_pb, entity, auto_id_entities):
     auto_id = entity.key.is_partial
 
     key_pb = entity.key.to_protobuf()
-    key_pb = helpers._prepare_key_for_request(key_pb)
```
Comment (tseaver) [inline gcloud/datastore/batch.py]: Yay!

File: gcloud/datastore/test_connection.py [hunk 3]
```diff
@@ -837,7 +837,7 @@ def test_allocate_ids_non_empty(self):
         request.ParseFromString(cw['body'])
         self.assertEqual(len(request.key), len(before_key_pbs))
         for key_before, key_after in zip(before_key_pbs, request.key):
-            _compare_key_pb_after_request(self, key_before, key_after)
```
Comment (tseaver) [inline gcloud/datastore/test_connection.py]: Yay!

**Comment by tseaver** [comment]
LGTM.

**Comment by dhermes** [comment]
@tseaver We can't / shouldn't merge this until the discussion above is resolved.

**Comment by tseaver** [comment]
Until this lands, or the back-end quits returning keys with `s~` prefixes, I can't DTRT in #552 (fixing #447), unless I re-inject the "unmangling" logic we've already cleaned out elsewhere.

File: gcloud/datastore/__init__.py [hunk 4]
```diff
@@ -69,6 +69,50 @@
 _DATASET_ENV_VAR_NAME = 'GCLOUD_DATASET_ID'
 
 
+def _find_true_dataset_id(dataset_id, connection=None):
```
Comment (tseaver) [inline gcloud/datastore/__init__.py]: We likely need to make this method public, to allow applications which don't use the inferred dataset ID to normalize their project ID -> dataset ID.

Comment (dhermes) [inline gcloud/datastore/__init__.py]: I'd prefer they just use `s~foo`, but hopefully none of that will matter once the backend is updated.

**Comment by tseaver** [comment]
Ping me if more review is needed.

**Comment by dhermes** [comment]
Rebased and squashed into single commit (had to hack around after rebasing).

File: gcloud/datastore/__init__.py (hunk 4 above)
Comment (tseaver) [inline gcloud/datastore/__init__.py]: ISTM that making people know about `s-foo` is an anti-goal, even in the short term.

Comment (dhermes) [inline gcloud/datastore/__init__.py]: RE: "know about", if they insert an entity, that will be the dataset ID returned by a `get`, hence the use of the fuzzy comparison in the regression test.

Also, by leaving this out we have previously accidentally introduced bugs like #447

**Comment by dhermes** [comment]
I'm closing this (since it is so out of date) and moving to discussion / tracking in #821.

Part of it was merged in #792 as well.

**Pull request closed without merging**